[docs] Discovery service (#1522)
Azure/iotedge-lorawan-starterkit: docs/user-guide/lns-discovery.md

# LNS discovery

> The LNS protocol specifies that the first request an LBS makes is to the `/router-info` discovery endpoint. The LNS from the starter kit implements the `/router-info` endpoint and redirects any requests to that endpoint forward to the `/router-data` endpoint on the same LNS.

Reviewer: ```suggestion
The LNS protocol specifies that the first request an LBS makes is to the `/router-info` discovery endpoint. The LNS from the starter kit implements the `/router-info` endpoint and redirects any requests to that endpoint to the `/router-data` endpoint on the same LNS.
```

> In addition to the discovery endpoint on the LNS itself, we also support a standalone discovery service, which you can use to distribute LBS connections to different LNS. The discovery service is a .NET application, and its implementation is part of the `LoRaWan.NetworkServerDiscovery` project.

Reviewer: adding a link to the LoRaWan.NetworkServerDiscovery source?

Reviewer: ```suggestion
In addition to the discovery endpoint on the LNS itself, we also support a standalone discovery service, which you can use to dynamically distribute LBS connections to different LNS, ensuring failover of LBS to antother LNS in case the current LNS is experiencing downtime.. The discovery service is a .NET application, and its implementation is part of the `LoRaWan.NetworkServerDiscovery` project.
```

The discovery service relies on configuration from IoT Hub module twins to associate each LBS with a set of LNS to which it may connect to. For this the configuration needs to take into account possible network boundaries - by defining these networks you can control which LBS should connect to which LNS. For this you need to configure your LBS twin in IoT Hub as follows:

```json
{
  ...,
  "tags": {
    "network": "<network-id>"
  },
  ...
}
```

> The network ID may only consist of alphanumeric characters. After you configured the LBS, you need to also configure your LNS with the same network. The LNS twin furthermore needs to contain the IP/DNS of the host, hence the twin of the LNS *module* should look like:

Reviewer 2: Let's highlight this as a warning. It might pass un-noticed

Author: Done in 2a0946c18c18de34c57da15dbdaa9ec8bf60622a

```json
{
  ...,
  "tags": {
    "network": "<network-id>"
  },
  ...,
  "properties": {
    "desired": {
      "hostAddress": "wss://mylns:5001",
      ...
    }
  }
}
```

> Keep in mind that the `hostAddress` value must include the scheme and the host name. In case you omit the port, the defaults of the protocol will be used.

Reviewer 2: What about asking to always specify the port too?
Default of the protocol are useless in our starter kit

Author: It would be stricter, but it would also require code changes. I would argue that if they change the port mappings that we use per default in our starter kit, it might still be useful to keep it.

!!! note
    The discovery endpoint distributes LBS to LNS using a round-robin distribution mechanism. It will always try the same LNS per LBS first. Hence, it can be used for an active/passive scenario, but not to distribute load more evenly across different LNS.

!!! note

>     The configuration values of the network name of a station, respectively of the set of LNS in a given network, are cached by the discovery endpoint for six hours.

Reviewer: ```suggestion
    The configuration values of the network name of a station, respectively of the set of LNS in a given network, are cached by the discovery endpoint for 6 hours.
```

>     If you updated the configuration, make sure to restart the discovery service to ensure that the cache is refreshed.

Reviewer: ```suggestion
    If you updated the configuration and want it to take effect sooner, make sure to restart the discovery service to ensure that the cache is refreshed.
```

You can choose to deploy the discovery service either on-prem or in the cloud. For both deployment strategies, you can configure the following behavior:

> - The log level can be configured by setting the `Logging__LogLevel__Default` environment variable. For more fine-grained configuration of the console log level, refer to [Logging in .NET Core and ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-6.0#set-log-level-by-command-line-environment-variables-and-other-configuration).

Reviewer: it would be great to have a table with configuration settings that are supported, a description, supported values and the default we apply, if not specified.

Author: Good idea, done in 3debf30343125d264b71101fe0e66cb71967a5fe

- You can use Application Insights by setting the `APPINSIGHTS_INSTRUMENTATIONKEY` environment variable. The log levels for Application Insights can be configured similarly to the default log levels, but by using the environment variables `Logging__ApplicationInsights__LogLevel__<Default|...>`.
- You can configure on which URLs the service is listening by setting the environment variable `ASPNETCORE_URLS`.

## Deployment in Azure

We recommend that you use an Azure App Service to run the discovery service in the cloud. You can follow the [App Service Quickstart](https://docs.microsoft.com/en-us/azure/app-service/quickstart-dotnetcore?tabs=net60&pivots=development-environment-cli) for instructions how to deploy an ASP.NET Core app in an Azure App Service. When using an App Service, make sure that you:

- [Enable the system assigned managed identity](https://docs.microsoft.com/en-us/azure/app-service/overview-managed-identity?tabs=portal%2Chttp#add-a-system-assigned-identity)
- Assign a role to the system assigned identity which allows it to query the IoT Hub twins, e.g. IoT Hub Twin Contributor.

> - Specify the `IotHubHostName` environment variable, such that it points to your IoT Hub.

Reviewer: should also go in the configurations table.

Author: Done in 3debf30343125d264b71101fe0e66cb71967a5fe

> - Assign a role to the system assigned identity which allows it to query the IoT Hub twins, e.g. IoT Hub Twin Contributor.

Reviewer 2: ```suggestion
- Assign a role to the system assigned identity which allows it to execute the `Microsoft.Devices/IotHubs/twins/read` data action for reading the IoT Hub twins. Built-in `IoT Hub Twin Contributor` role allows that.
```

!!! note
    The configuration values of the network name of a station, respectively of the set of LNS in a given network, are cached by the discovery endpoint for 6 hours.
    If you updated the configuration and want it to take effect sooner, make sure to restart the discovery service to ensure that the cache is refreshed.

You can choose to deploy the discovery service either on-prem or in the cloud. For both deployment strategies, you can configure the following behavior:

## Configuration

You can configure the following behavior of the LNS discovery service.

| Environment variable | Description | Deployment type |
| ---                  | ---         | ---             |
| `Logging__LogLevel__Default` | Configures the default log level. For more fine-grained configuration of the console log level, refer to [Logging in .NET Core and ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-6.0#set-log-level-by-command-line-environment-variables-and-other-configuration) | All |
| `APPINSIGHTS_INSTRUMENTATIONKEY` | Sends telemetry data to Application Insights | All |

> | `Logging__ApplicationInsights__LogLevel__<Default|..>` | Configures the Application Insights log levels | All |

Reviewer 3: [nit] I think this might not render correctly in a table and whaterver is after '|' would end up in a separate column. The same goes for line 63.

Author: Not sure I can reproduce, when rendering locally it looks correct to me:

![image](https://user-images.githubusercontent.com/22341213/155284598-27557800-b9ad-42d0-b55c-0f5bae9ae381.png)

Which renderer are you using?

Reviewer 3: It was split into different columns here in the PR, when selecting the preview for the file. But since it's rendering as expected in the documentation page, then it's not a problem.
